[JSC] Add more assertions to help diagnose a crash below JSModuleRecord::instantiateDeclarations

hyjorc1 opened the pull request:
#### 272586c8729f97253802bfba634221c90eb41efc
<pre>
[JSC] Add more assertions to help diagnose a crash below JSModuleRecord::instantiateDeclarations
<a href="https://bugs.webkit.org/show_bug.cgi?id=257896">https://bugs.webkit.org/show_bug.cgi?id=257896</a>
rdar://110339520

Reviewed by NOBODY (OOPS!).

Previously, we landed a patch [1] to mitigate an unexplained crash
due to an undefined JSModuleEnvironment* in JSModuleRecord::instantiateDeclarations.
Initially, we suspected the crash due to an undefined entry gotten from
the dependencies map. With info collected by the follow-up patch [2], we found that the
entry is not undefined but with an undefined property `module`. This might
be due to the potential module instantiation failure before filling the registered
entry in requestInstantiate of ModuleLoader.js. To support our hypothesis,
add more assertions for more info in AbstractModuleRecord::hostResolveImportedModule.

[1] <a href="https://commits.webkit.org/262608@main">https://commits.webkit.org/262608@main</a>
[2] <a href="https://commits.webkit.org/262687@main">https://commits.webkit.org/262687@main</a>

* Source/JavaScriptCore/runtime/AbstractModuleRecord.cpp:
(JSC::AbstractModuleRecord::hostResolveImportedModule):
* Source/JavaScriptCore/runtime/JSCJSValueInlines.h:
(JSC::JSValue::getPropertySlot const):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/272586c8729f97253802bfba634221c90eb41efc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9929 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10176 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10426 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11581 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9626 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12162 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10128 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12571 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10083 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10879 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8413 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11964 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8186 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9004 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16344 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/8417 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9284 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9153 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12427 "11 api tests failed or timed out") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/9429 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9655 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7844 "4 failures") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/10072 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8814 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2580 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13040 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/10342 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9429 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2562 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/runtime/JSModuleRecord.cpp:
```diff
@@ -155,12 +155,8 @@ void JSModuleRecord::instantiateDeclarations(JSGlobalObject* globalObject, Modul
-            if (!vm.exceptionForInspection() || !vm.traps().maybeNeedHandling()) {
-                throwSyntaxError(globalObject, scope, makeString("Importing module '", String(importEntry.moduleRequest.impl()), "' is not found."));
-                return;
-            }
```
Why remove this check?  If an exception is already thrown, why override it here?

hyjorc1 committed:
[JSC] Add more assertions to help diagnose a crash below JSModuleRecord::instantiateDeclarations
https://bugs.webkit.org/show_bug.cgi?id=257896
rdar://110339520

Reviewed by NOBODY (OOPS!).

Previously, we landed a patch [1] to mitigate an unexplained crash
due to an undefined JSModuleEnvironment* in JSModuleRecord::instantiateDeclarations.
Initially, we suspected the crash due to an undefined entry gotten from
the dependencies map. With info collected by the follow-up patch [2], we found that the
entry is not undefined but with an undefined property `module`. This might
be due to the potential module instantiation failure before filling the registered
entry in requestInstantiate of ModuleLoader.js. To support our hypothesis,
add more assertions for more info in AbstractModuleRecord::hostResolveImportedModule.

[1] https://commits.webkit.org/262608@main
[2] https://commits.webkit.org/262687@main

* Source/JavaScriptCore/runtime/AbstractModuleRecord.cpp:
(JSC::AbstractModuleRecord::hostResolveImportedModule):
* Source/JavaScriptCore/runtime/JSCJSValueInlines.h:
(JSC::JSValue::getPropertySlot const):

Constellation commented on Source/JavaScriptCore/runtime/JSCJSValueInlines.h:
```diff
@@ -1069,8 +1069,11 @@ ALWAYS_INLINE bool JSValue::getPropertySlot(JSGlobalObject* globalObject, Proper
         object = synthesizePrototype(globalObject);
         EXCEPTION_ASSERT(!!scope.exception() == !object);
-        if (UNLIKELY(!object))
+        if (UNLIKELY(!object)) {
+            VM& vm = getVM(globalObject);
+            RELEASE_ASSERT(vm.exceptionForInspection() && vm.traps().maybeNeedHandling(), vm.traps().maybeNeedHandling(), vm.exceptionForInspection(), JSValue::encode(*this));
             return false;
```
This is one of the very hot path. Let's not add a cost to that, and make it scoped to AbstractModuleRecord.cpp.
